### ISSUE TITLE
[Question] Sort exception message by references

### ISSUE THREAD

**Issue by obstschale** [Respect/Validation#951]
I know that `$exception->getMessages()` will return an array with all validation errors. But is there a way to get all errors as array where the errors are sorted by the reference?

Example:

```php
$params = [ 'songid' => 'A', 'arrangementid' => 'B' ];

try {
    $fileValidator = Validator::key('songid', Validator::numeric()->length(1,32))
                              ->key('arrangementid', Validator::numeric()->length(2, 4));

    $fileValidator->assert($params);
} catch (ValidationException $e) {
   $errors =  $e->getMessages();
}
```

```
// $errors will be:
[
        "songid must be numeric",
        "All of the required rules must pass for arrangementid",
        "arrangementid must be numeric",
        "arrangementid must have a length between 2 and 4"
]
```

But I would like to have something like this:

```
[
    'songid' => [
            "songid must be numeric",
    ],
    'arrangementid' => [
        "All of the required rules must pass for arrangementid",
        "arrangementid must be numeric",
        "arrangementid must have a length between 2 and 4"
    ]
]
```

Is this possible?

**Comment by nickl-** [comment]
Have you tried `$e->getIterator()` which will collect the exceptions for you. On each exception you will have `getParams()`, `getName()` and of course `getMessage()` which should provide you with all the detail you need.

**Comment by nealoke** [comment]
@nickl- when I try this I get the following:

`Fatal error:  Allowed memory size of 134217728 bytes exhausted (tried to allocate 65015808 bytes) in C:\........ on line xx`

**Comment by nickl-** [comment]
Ouch! That's not right...I know the data you are looking for is in there somewhere. It is going to take me a little while before I am free again to have a look.

**Comment by henriquemoody** [comment]
@nealoke, can you send out the exact piece of code you're executing? This library is very lightweight to have this kind of memory problem, it may be possible but I'd like to know exactly what's happening.

**Comment by nealoke** [comment]
@henriquemoody sure, below is the example I use:

```
    try {
        return v::keySet(
            v::key("property", v::intVal()),
            v::key("property", v::intVal()),
            v::key("property", v::intVal()),
            v::key("property", v::intVal()),
            v::key("property", v::intVal()),
        )->assert($input);
    }

    catch(NestedValidationException $exception) {
        $errors = $exception->getIterator(); 
        print_r($errors);
    }
```

**Comment by nickl-** [comment]
The idea is to iterate through the iterator, print_r might be reaching for more than we need which is why you are getting the stack overflow.

something on the lines of:
```php
catch (NestedValidationException $exception) {
        foreach($exception->getIterator() as $error)
            echo( $error->getName() . "  " .  $error-> getMessage());
}
```

Not sure if this even parses I can't test it right now but it makes sense right?

**Comment by henriquemoody** [comment]
That's probably will be fixed by #957

@nealoke, can you check if is that what you want? https://github.com/henriquemoody/Validation/blob/exception/tests/integration/get_messages.phpt

**Comment by nealoke** [comment]
@henriquemoody that is exactly how I would like it :). Any idea on when this will be merged? 😄

**Comment by henriquemoody** [comment]
I'm trying to merge it this week, but the release depends on #921

**Comment by nealoke** [comment]
@henriquemoody cool, I hope that it doesn't take too long 😄 . I'm not sure how I can help also :/

**Comment by henriquemoody** [comment]
@nealoke You can pick any of the rules listed on #921 and apply the contribution guidelines, just as #902 does. 😄

**Comment by henriquemoody** [comment]
Fixed by #957.

I've changed how `getMessages()`, check out [get_messages.phpt](https://github.com/Respect/Validation/blob/master/tests/integration/get_messages.phpt) and [get_messages_with_replacements.phpt](https://github.com/Respect/Validation/blob/master/tests/integration/get_messages_with_replacements.phpt) to see how it works.

It will be released with the version 2.0. If any of you want to speed things up take a look at #921!

Thanks for reporting! :panda_face:

**Comment by obstschale** [comment]
That's great :+1: Thank you.

If you wondered, I watched this thread silently all the time ^^
